File: tree/binary_tree_level_order_traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        //recursively?

        List<List<Integer>> ans = new LinkedList<List<Integer>>(); 
        List<Integer> curr_level; 
        Queue<TreeNode> q = new LinkedList<TreeNode>();
        if(root == null){
            return ans; 
        }
        q.add(root);
        
        TreeNode top; 
        while(!q.isEmpty() ){
            int size = q.size();
            curr_level = new LinkedList<Integer>();

            for(int i = 0 ; i < size; i++){
                //ss
                top = q.element();
                q.remove(); //pop 
                if(top.left != null){
                    q.add(top.left);
                }
                if(top.right != null){
                    q.add(top.right);
                }
                curr_level.add(top.val); //dequeue push elements into a deque from the back

            }
             ans.add(curr_level); //dequeu
        }
                    return ans; 

    }
   
    
}

File: easy/tree/binary_tree_level_order_traversal.cpp
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (root == nullptr) return {};
        
        queue<TreeNode*> q;
        vector<int> curr_level;
        vector<vector<int>> result;
        
        q.push(root);
        
        while (!q.empty()) {
            int size = q.size();
            for (int i = 0; i < size; ++i) {
                auto top = q.front();
                q.pop();
                if (top->left) q.push(top->left);
                if (top->right) q.push(top->right);
                curr_level.push_back(top->val);
            }
            result.push_back(curr_level);
            curr_level.clear();
        }
        
        return result;
    }
};